### PR TITLE
Coverage density raised to 100%  [issue #3207]

### LINKED ISSUE #3207: Increase density code test coverage
## Is your feature request related to a problem? ##
The current density code only has 82.04% test coverage (see [here](https://codecov.io/gh/MDAnalysis/mdanalysis/src/develop/package/MDAnalysis/analysis/density.py) )
The general aim is that we should get 100% coverage for everything.

## Describe the solution you'd like ##
Tests to cover the uncovered areas

## Additional context
This is an issue aimed at providing a simple issue for first time contributing GSoC students to work on.

### PR DESCRIPTION
**Pull request by ojeda-e**
Fixes #3207 
Density code test coverage raised to 100%.

Changes made in this Pull Request:
- Test coverage added in lines
-- 771-774 
-- 738-739
-- 754-755
-- 803-808
-- 810-869 (`convert_density`).

(No Docs needed)

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Review by PicoCentauri — APPROVED** [review]
LGTM 🚀

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG:26]:
```suggestion
  * Test coverage of the density code increased to 100% (Issue #3207) 
```

Sorry 😅 I know it's pedantic.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:116]: Sorry it's late and I probably am missing what we're doing here but shouldn't we be checking that the units have been set properly?

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:132]: As above

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:135]: What are we testing here?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Sorry I'm probably missing something really simple so please feel free to tell me I'm wrong.

It looks like we are calling convert methods but not checking their return? (i.e. I'm not actually sure what we are testing except that the code paths don't crash).

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_density.py:156]: Similar to Irfan's other comments--this just produces two strings so doesn't do much apart from flushing code paths. Making some kind of assertion about the contents of the strings seems like it might be in order for this case?

**Comment by ojeda-e** [comment]
Hi @IAlibay, sorry for the long delay.  

I think I changed the code as requested:

- Function `test_check_set_unit_density_none` added. 
Coverage raised from 82 -> 84% in lines 757-758.

- Function `test_parameters_isdensity` added. 
Coverage raised from 84 -> 86% in lines 789-792.

- Function `test_check_convert_density_grid_not_density` added. 
Coverage raised from 86 -> 89% in lines 869-871.

- Function `test_check_convert_density_value_error` added. 
Coverage raised from 89 -> 92% in lines 874-880.

- Function `test_check_convert_density_units_same_density_units` and
`test_check_convert_density_units_density` added. 
Coverage raised from 92 -> 93% in lines 872 and 881.

- Function `test_convert_length_same_length_units` added. 
Coverage raised from 93 -> 95% in line 822.

- Function `test_parameters_is_Density` added. 
Coverage raised from 95 -> 97% in line 883.

-Function `test_check_convert_length` added.
Coverage raised from 97 -> 100%; lines 824-827

About the last function, `test_check_convert_length`, I think you may want me to add an assertion for self.edges. I tried using `assert_almost_equal` for the two bumpy arrays, however, it didn't work. I got a warning about defining `dtype=object`. If you want that line added,I would like to have some suggestions of how to approach it from your side.
   
Thanks.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:155]: Rather than just checking that the units have been set. It would be good to also check that numerically the conversion has been made in a way that would be expected.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:170]: Same as above, we probably should be checking that the edges, origin, etc... are all as expected?

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:173]: Doesn't `convert_length` already call `_update` if it's not the same units?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Sorry I don't have time for a full review today. Here are some quick comments.
Where possible we should ensure that the right numerical operations have been done on the density.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: In the previous submission I had a question regarding the `numpy_array_almost_equal` to check the elements you mentioned are the same since I was getting the following error:
`numpy.VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.`
According to the documentation, `numpy_array_almost_equal` should work for numpy arrays. I got confused and was asking for some guide to approach this.

I think I fixed it by comparing element by element using `zip` as you can see for edges in `test_check_convert_density` and `test_check_convert_length` of the most recent commit (lines 168-169, and 175-176) . Can you please confirm that's a valid approach before adding to origin, etc?
Thanks.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: Thanks, you're correct. I rewrote all the test for `__repr__`.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:172]: Please double check, but I'm pretty sure this is a shallow copy, so it won't work as a comparison if you're modifying D in place (same for the test above).

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: If you can give us a example of what you were trying to use the `assert_almost_equal` on (note, if I remember correctly it's best to use that over `array_almost_equal` @lilyminium had some explanation about this somewhere), I can probably offer some insight into what's going on.

For the `grid`, `origin`, and `delta` attributes, assert_almost_equal should work as-is.

For the `edges` attribute you'll probably have to iterate (I can't remember if there's a better way to deal with that kind of list of arrays), something like what you're doing would work.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:159]: re-echoing Tyler's comment here. The `__repr__` is essentially a means of allowing a class to have a printable string.

Here I would create a density with and without `isDensity` and then do:

assert str(density_grid) = '<Density density with insert_shape_here bins>'
assert str(not_density_grid) = '<Density histogram with insert_shape_here bins>'

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_density.py:148]: Note: it would be more than reasonable [if not even suggested] to have this test as part of the same test where you compare the grid value change (same for the convert_length stuff).

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: An example of the behaviour is below
```
    def test_check_fail(self, D, h_and_edges):
        h, edges = h_and_edges
        assert_almost_equal(edges, D.edges)
    
    def test_check_no_fail(self, D, h_and_edges):
        h, edges = h_and_edges
        for i, j in zip(edges, D.edges):
            assert_almost_equal(i, j)
```
The error I get is
```
E       numpy.VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.

../../../opt/anaconda3/envs/mdanalysis-dev/lib/python3.7/site-packages/numpy/core/_asarray.py:102: VisibleDeprecationWarning
```

The assertion is intended for the `edges`, which is a numpy array. [The documentation](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html) gives an example of comparison of two numpy arrays. 
Printing `edges` and `D.edges` reads as
```
[array([ 0.        ,  3.33333333,  6.66666667, 10.        ]), array([ 0. ,  2.5,  5. ,  7.5, 10. ]), array([ 0.,  2.,  4.,  6.,  8., 10.])]
[array([ 0.        ,  3.33333333,  6.66666667, 10.        ]), array([ 0. ,  2.5,  5. ,  7.5, 10. ]), array([ 0.,  2.,  4.,  6.,  8., 10.])]

```
Hence my question if the approach by using `zip` to compare individual elements is valid. I get that in this case `edges` is list of arrays. Which I eventually could transform to `np.array(edges)`, but even in that case `assert_almost_equal` fails and prints the same error.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py:172]: Is this better? (`grid` is missing, though)

```
    def test_check_convert_length(self, D, h_and_edges, bins):
        midpoints = [0.5*(bin[:-1] + bin[1:]) for bin in bins]
        deltas = np.array([self.Lmax])/np.array(self.nbins)
        origin = [mp[0] for mp in midpoints]
        h, edges = h_and_edges
        new_unit = 'nm'
        D.convert_length(new_unit)

        for prev_edge, conv_edge in zip(edges, D.edges):
            assert_almost_equal(prev_edge*0.1,  conv_edge)
        
        for prev_origin, conv_origin in zip(origin, D.origin):
            assert_almost_equal(prev_origin*0.1,  conv_origin)
        
        for prev_delta, conv_delta in zip(deltas, D.delta):
            assert_almost_equal(prev_delta*0.1,  conv_delta)
```

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py:148]: Sure, I considered to move it to same function that check convert_density but was hesitating. Thanks.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: Yes, my bad, now it's gone.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py:159]: Ok, no, I would have never seen that myself. Thanks.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: The zip approach is good: the elements of the list are simply arrays of different lengths; I can't think of a more sensible way to do the tests than what you did.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [reply testsuite/MDAnalysisTests/analysis/test_density.py:148]: See my comment with `deepcopy` on how you could also check that the values changed:
```python
...
assert D.units['density'] ==  unit
assert_almost_equal(D.grid, 10**3 * D_orig.grid)   # or write 1e3 * ... but I wanted to make the factor clear
```

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_density.py:154]: pep8 spaces
```suggestion
        unit = 'nm'
```

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_density.py:153]: slightly more readable if you put the reference on the RHS
```suggestion
        assert D.units['length'] == unit
```

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_density.py:156]: switch LHS with RHS

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_density.py:154]: When you switch to `nm`, you're not testing "same length unit" any more, are you? Then it would be better to make it a separate test.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_density.py:151]: Also test the _value_ of the grid array.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_density.py:158]: rename `test_repr`.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (orbeckst) [reply testsuite/MDAnalysisTests/analysis/test_density.py:172]: You can use `copy.deepcopy()` before changing the Density.

`Density` could really have a `Density.copy()` method to do it by itself, but that's for another day.

**Review by orbeckst — CHANGES_REQUESTED** [review]
As stated by previous reviewers, we want to test _values_ in addition to metadata changes. One possible approach is to make deep copies (see comments). Please address the outstanding comments and then we can merge this — thank you!!

**Comment by orbeckst** [comment]
@ojeda-e I merged the latest develop (in CHANGELOG I added your name to 2.1.0 but I removed the entry itself because users don't care about test coverage in the CHANGELOG). If you could address the remaining comments then we could probably get this in quickly.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: Thanks. Separate test added.

File: package/CHANGELOG
Comment (orbeckst) [inline package/CHANGELOG]: The CHANGELOG got messed up, possibly by force pushing.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Code looks good, just CHANGELOG got messed up and you need to add entries under 2.1.0. Perhaps pull develop and then rebase your branch onto the latest develop?

**Comment by ojeda-e** [comment]
Sorry about it, @orbeckst.
I think it's fixed now although I don't fully understand the change shown in the [last line](https://github.com/MDAnalysis/mdanalysis/pull/3249/files#diff-4ac8d9ef37778aec14402a8cba558978f084f8359a6fbb7f494c7be8b8e6c5ccR2770) of the CHANGELOG.
The file from the `develop` branch has no line at the end. 
Thanks.

**Comment by orbeckst** [comment]
Now you're missing from CHANGELOG, you should be listed as a 2.1.0 author. Do you want to update the CHANGELOG before I merge?

Just add the new line ending in the last line in the CHANGELOG back, it seems your editor removed it eagerly in 
57d66ce ; the smaller the diff the better and then at least someone else's editor will not auto-add it back. Teach your tools to only change things that _you_ want to edit (they don't always know better ;-) ).

**Review by orbeckst — APPROVED** [review]

**Comment by orbeckst** [comment]
@IAlibay could you please check here if all of your comments have been addressed? Thanks!

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py]: The test logic here isn't clear to me.

1) You start with units of {'length': 'A', 'density': 'A^{-3}'}
2) Then you set units to None again
3) Then you delete the density units
4) Then you test that passing an empty dictionary will yield units of None for 'density'

The only thing that properly gets tested here is number 4 in that list (3 is kinda meant to test 4). What's the point for actions 1 and 2, why not just delete the units from the start? If the point is to check what happens when you pass None, then should there be a second assert?

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:149]: Probably should add an assert here that checks that the unit was set properly (that would match the same test behaviour as the test just above).

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py]: str(D) should just call `__repr__` in the background right? I don't _think_ these two calls are needed?

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:381]: The visual indentation here is probably worse? Is this the result of an auto-formatter? Personally I'd prefer we stick with the previous formatting for this line.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:238]: IMHO this is visually worse than what it was before. If it's to avoid PEP8 issues, I'd at least space indent the arguments here.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py:207]: The indentation here is no longer consistent with the rest of the dictionary, unless there's a good reason, I'd just stick with what it was?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of quick comments.

**Comment by orbeckst** [comment]
@ojeda-e can I motivate you to address the requested changes to get the PR done and into the 2.1.0 release?

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py:207]: Thanks for pointing this out, indentation was unintended. Fixed here and lines below.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: Thanks, _I think_ I followed your suggestions and change the test:
1. deleted units, 
2. assert units,
3. assert check_set_units.

**Comment by ojeda-e** [comment]
Thanks for the motivation @orbeckst. ⚡ 
@IAlibay, I hope I addressed all your comments, sent a request to have your review.

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: This is going to have to spillover to the next line next time someone adds an entry - we should just break it into two lines now to save time later.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py]: I might be missing the point of what this is doing here.
`_check_set_unit` has no return right? So asserting None is doesn't do anything?
Shouldn't you be asserting that `D.units['density'] is None`? (note that `is` is better than `==` for `None` checking).

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py]: Which bit of the code is being tested by passing the same units to itself?

Also - same issue as previous comment, unless I'm missing something obvious `_check_set_unit` has no return so you need to assert on `D.units`.

**Review by IAlibay — CHANGES_REQUESTED** [review]
Apologies, couple more things I didn't catch on to before it seems.

**Comment by IAlibay** [comment]
@ojeda-e just a reminder that there are two final review queries to address on this PR.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: Yes, sorry about that duplicated `test_check_set_unit_density_none`, this one was supposed to be deleted. The unit test is intended to cover lines 772-773 in `density.py`.

**Comment by ojeda-e** [comment]
My apologies for making the progress in this PR so stupidly slow @IAlibay, also thanks for spotting the duplicated unit test. Pushed the changes you pointed out. Thanks.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py]: Sorry I'm going to have to query this again, I don't see how asserting `None` on a function that returns nothing does anything else but flush the code path?

Is this just checking lines 755-756?

If so this is what I would do here:

```suggestion
        D._check_set_unit(units)
        assert D.units['density'] is None
```

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_density.py]: nitpicking but is passing a unit necessary here? Shouldn't kwarg default also trigger this error?

**Review by IAlibay — CHANGES_REQUESTED** [review]
I think we're circling around the same test, I've provided a suggestion for what I _think_ `test_check_set_unit_density_none` should be doing instead. Once dealt with, this PR will be good to go.

File: testsuite/MDAnalysisTests/analysis/test_density.py
Comment (ojeda-e) [reply testsuite/MDAnalysisTests/analysis/test_density.py]: Yes, you're correct, thanks for catching this.

**Comment by ojeda-e** [comment]
Thanks @IAlibay for clarifying.

Before I submit the changes, I would like to explain what I did and check if I am doing something wrong once again.
There are two cases in `density.py` that involve the key `density` in units dict.

The first one is when `density` is a key in units dict, but has value `None` (line 756), which can be tested with:
```python3
    def test_check_set_density_none(self, D1):
        units = {'density': None}
        D1._check_set_unit(units)
        assert D1.units['density'] is None
```

The second one is when `density` is not a key in units dict (lines 773). To test these lines, I added 
```python3
    def test_check_set_density_not_in_units(self, D1):
        del D1.units['density']
        D1._check_set_unit({})
        assert D1.units['density'] is None
```
Sounds/looks correct?

**Comment by IAlibay** [comment]
> Thanks @IAlibay for clarifying.
> 
> Before I submit the changes, I would like to explain what I did and check if I am doing something wrong once again. There are two cases in `density.py` that involve the key `density` in units dict.
> 
> The first one is when `density` is a key in units dict, but has value `None` (line 756), which can be tested with:
> 
> ```python
>     def test_check_set_density_none(self, D1):
>         units = {'density': None}
>         D1._check_set_unit(units)
>         assert D1.units['density'] is None
> ```
> 
> The second one is when `density` is not a key in units dict (lines 773). To test these lines, I added
> 
> ```python
>     def test_check_set_density_not_in_units(self, D1):
>         del D1.units['density']
>         D1._check_set_unit({})
>         assert D1.units['density'] is None
> ```
> 
> Sounds/looks correct?

Yup, those two tests sound perfect! Thanks @ojeda-e !

**Review by IAlibay — APPROVED** [review]
lgtm (assuming CI returns green)

**Comment by IAlibay** [comment]
Codecov reports 100% coverage in density.py, thanks @ojeda-e !

**Pull request merged**
